### PR TITLE
Upload releases to s3

### PR DESCRIPTION
**Pull request by joshaber**
Upload releases to a bucket separate from Releases.

This is a step toward fixing #9658.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
   "dependencies": {
     "async": "0.2.6",
     "atom-keymap": "^6.1.1",
+    "aws-sdk": "^2.2.18",
```
Comment (joshaber) [inline package.json]: Given that we only need for this CI, should we install it in `cibuild` (or some such) instead? @atom/feedback

Comment (maxbrunsfeld) [inline package.json]: What we've been doing is putting these dependencies into `build/package.json`.

Comment (joshaber) [inline package.json]: :+1:

**Comment by joshaber** [comment]
I think this is ready :pray:

@maxbrunsfeld I'd love your :eyes: on this to make sure I didn't mess up releases.

@thedaniel Do you think this makes sense? Does this seem ok for the atom.io side of things?

**Comment by maxbrunsfeld** [comment]
:+1: This looks good to me @joshaber, though I'm still not totally fully fluent in these scripts myself. I think we should go for it.

**Comment by joshaber** [comment]
@thedaniel @maxbrunsfeld I set the object to not be cached (https://github.com/atom/atom/commit/7b976bc689e479995ca1ce40a5f2cc357aa5984a) and it seems to work as we'd expected. **But** I accidentally stumbled upon what I think is an even better solution: we can [configure CloudFront](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/QueryStringParameters.html) to take query string parameters into account with respect to caching.

So if atom.io throws the asset's last modified date onto the download URL, we should get consistent caching and the right download regardless of how many times the file is changed in place :metal:

**Comment by joshaber** [comment]
(I hit submit a bit prematurely there.)

So given that we probably weren't gonna ship this until next week anyways, I'm gonna take the time to reconfigure CloudFront and then allow caching again.

**Comment by joshaber** [comment]
@atom/feedback I believe this is ready :pray: 

My current plan:

1. Merge into `master`.
2. Cherry-pick into `stable` and `beta`.
3. Ensure the build binaries are uploaded for beta.
4. Upload the current release build binaries manually.

**Comment by lee-dohm** [comment]
It looks good to me. I don't see any glaring issues. Though I also am not terribly familiar with these scripts.

File: build/tasks/publish-build-task.coffee [hunk 2]
```diff
@@ -221,15 +222,43 @@ uploadAssets = (release, buildDir, assets, callback) ->
...
+
+    key = "releases/#{release.tag_name}/#{assetName}"
+    uploadParams =
+      Bucket: s3Bucket
+      ACL: 'public-read'
+      Key: key
+      Body: fs.createReadStream(assetPath)
+    s3.upload uploadParams, (error, data) ->
+      if error?
+        console.log("Upload release asset #{assetName} to S3 failed", error)
```
Comment (kevinsawicki) [inline build/tasks/publish-build-task.coffee]: Should this use `logError`?

Comment (joshaber) [inline build/tasks/publish-build-task.coffee]: Yup, good catch :+1:

**Comment by joshaber** [comment]
Dunno what's up with Travis.

**Pull request merged**